### PR TITLE
Make temp tables suffix length flexible

### LINKED ISSUE #299: Should be able to create more than 1000 temporary tables
Hi maintainers,

When we use modes other than `INSERT_DIRECT` and `MERGE_DIRECT` in embulk-output-jdbc (or plugins that inherit from embulk-output-jdbc), there is a restriction that we cannot create more than 1000 temporary tables as follows.

ref. https://github.com/embulk/embulk-output-jdbc/blob/7a433dc/embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java#L698-L704

This will cause problems when the number of tasks exceeds 1000, such as when there are more than 1000 input files.

I think this restriction was designed to prevent the maximum number of table names from being exceeded, but I think the problem can be solved if the number of suffix digits can be specified as an option. What do you think?

### PR DESCRIPTION
**Pull request by civitaspo**
close #299

### CODE REVIEW

**Comment by dmikurube** [comment]
How about extracting that number-formatting part as a method, and adding some quick tests for that?

**Comment by civitaspo** [comment]
@dmikurube Thanks for pointing that out. I extracted the method and wrote tests for it.

File: embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java [hunk 1]
```diff
@@ -693,10 +694,11 @@ protected TableIdentifier buildIntermediateTableId(final JdbcOutputConnection co
                 {
                     intermTables = new ArrayList<>();
                     if (task.getMode().tempTablePerTask()) {
-                        String namePrefix = generateIntermediateTableNamePrefix(task.getActualTable().getTableName(), con, 3,
+                        int suffixLength = Math.max(1, String.valueOf(taskCount - 1).length());
```
Comment (dmikurube) [inline embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:697]: Sorry for nitpicking…, but this `suffixLength` calculation part could be in the `build…` method by passing `taskCount`? That would be more modular and would have better test coverage.

Comment (civitaspo) [reply embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:697]: @dmikurube This is because `suffixLength` is passed as an argument to the `genereateIntermediateTableNamePrefix` method. If testing this calculation, I'll have to create another method like `calculateSuffixLength`, is that ok?

Comment (civitaspo) [reply embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:697]: Or, is it better to create a separate implementation of `genereateIntermediateTableNamePrefix` that takes `taskCount` as an argument, and embed the `suffixLength` calculation in each of its methods and the `build...` method?

Comment (civitaspo) [reply embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java]: See. https://github.com/embulk/embulk-output-jdbc/pull/301/commits/02eeae77008a7c4f59a7d132aed0a8030bb81b2c

File: embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java [hunk 2]
```diff
@@ -760,6 +761,28 @@ private void dropTables() throws SQLException
...
+    }
+
+    @VisibleForTesting
+    static int calculateSuffixLength(int taskCount)
+    {
+        assert(taskCount >= 0);
+        return Math.max(1, String.valueOf(taskCount - 1).length());
+    }
+
+    protected String generateIntermediateTableNamePrefixByTaskCount(String baseTableName, JdbcOutputConnection con,
```
Comment (civitaspo) [inline embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:779]: I avoided writing tests for this method because I would have to make changes to parts that are not related to this PR. Since the internal implementation is time-dependent, it is impossible to write tests without first separating the dependency.

**Review by dmikurube — APPROVED** [review]
Ahh..., got it. `suffixLength` was used also in `generateIntermediateTableNamePrefix`. Sorry, my comment made the change complex.

I have another idea, but let me approve it now because the further change could be much bigger. I leave two options here up to you.

1. Go ahead as is. (Still better to have @hito4t's review.)
2. Stop predefining `namePrefix`, and build the entire table name every time only in `build…`.
  * I don't see so much benefit from predefining `namePrefix`. No big performance benefit, and less testable.

**Comment by civitaspo** [comment]
> I don't see so much benefit from predefining namePrefix. No big performance benefit, and less testable.

I agree with your opinion. But, I think it would be better to do this with a different PR because, as you say, the amount of changes would be a bit much to stop predefining `namePrefix`. So I would like to release this PR as it is.

**Review by hito4t — COMMENTED** [review]
@civitaspo @dmikurube 
How about defining `generateIntermediateTableNameFormat` method, which returns `namePrefix + "%0" + suffixLength + "d"` ?
We are able to get the intermediate table name by `String.format(tableNameFormat, taskIndex)`.
By that, we can avoid calculating `suffixLength` twice.

**Comment by civitaspo** [comment]
@hito4t

> How about defining `generateIntermediateTableNameFormat` method, which returns `namePrefix + "%0" + suffixLength + "d"` ?

I think this is a good idea. I agree with this way as it simplifies the implementation.

---

So I implemented this in efac12e. Could you review this?

File: embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java [hunk 3]
```diff
@@ -760,6 +761,22 @@ private void dropTables() throws SQLException
...
     }
 
+    @VisibleForTesting
+    static int calculateSuffixLength(int taskCount)
+    {
+        assert(taskCount >= 0);
+        return Math.max(1, String.valueOf(taskCount - 1).length());
+    }
+
+    protected String generateIntermediateTableNameFormat(String baseTableName, JdbcOutputConnection con,
```
Comment (civitaspo) [inline embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:771]: I avoided writing tests for this method because I would have to make changes to parts that are not related to this PR. Since the internal implementation is time-dependent, it is impossible to write tests without first separating the dependency.

File: embulk-output-jdbc/src/test/java/org/embulk/output/jdbc/TestJdbcOutputPlugin.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            try {
+                AbstractJdbcOutputPlugin.calculateSuffixLength(0);
+            } catch (java.lang.AssertionError e) {
+                assert(true);
+            }
+        }
+        {
+            int i = AbstractJdbcOutputPlugin.calculateSuffixLength(1);
+            assertEquals(1, i);
+        }
```
Comment (dmikurube) [inline embulk-output-jdbc/src/test/java/org/embulk/output/jdbc/TestJdbcOutputPlugin.java:29]: It was generating `"001"` for `1`, but it is going to generate `"1"` for `1`?

I thought:
* `001`, `002`, ..., `999`, `1000`, `1001`, ...

But is this actually :
* `1`, `2`, ..., `999`, `1000`, `1001`, ...

Comment (hito4t) [reply embulk-output-jdbc/src/test/java/org/embulk/output/jdbc/TestJdbcOutputPlugin.java:29]: @civitaspo @dmikurube 
If the minimum value for `suffixLength` is 3, we can maintain compatibility when `taskCount` is less than 1000.

Comment (civitaspo) [reply embulk-output-jdbc/src/test/java/org/embulk/output/jdbc/TestJdbcOutputPlugin.java:29]: @dmikurube @hito4t Oh, I'm sorry. I had unintentionally crushed backward compatibility. I'll fix it!

Comment (civitaspo) [reply embulk-output-jdbc/src/test/java/org/embulk/output/jdbc/TestJdbcOutputPlugin.java]: @dmikurube @hito4t Apply the review in https://github.com/civitaspo/embulk-output-jdbc/commit/c2ac8781b81b002b9be59f3ef405dee180c67bf6.

**Review by dmikurube — APPROVED** [review]
Still LGTM. :)

File: embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java [hunk 5]
```diff
@@ -760,6 +761,22 @@ private void dropTables() throws SQLException
         }
     }
 
+    @VisibleForTesting
+    static int calculateSuffixLength(int taskCount)
+    {
+        assert(taskCount >= 0);
+        return Math.max(3, String.valueOf(taskCount - 1).length());
```
Comment (hito4t) [inline embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java:768]: @civitaspo 
It will be more readable if there is a comment about the reason for the magic number `3` (for backward compatilibity).

A commreason

**Review by hito4t — CHANGES_REQUESTED** [review]

File: embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java (hunk 5 above)
Comment (civitaspo) [reply embulk-output-jdbc/src/main/java/org/embulk/output/jdbc/AbstractJdbcOutputPlugin.java]: Oops, I'm sorry. You're right. I added a commit https://github.com/embulk/embulk-output-jdbc/pull/301/commits/89a63aabee57e4bc105538a639e68c3243465fc0. Please review again.

**Review by hito4t — APPROVED** [review]
LGTM!

**Comment by civitaspo** [comment]
@hito4t san or @dmikurube san Could you merge this PR and release it as a new version?

**Comment by dmikurube** [comment]
Thanks for your contribution! Merging.

**Comment by civitaspo** [comment]
😄 😄 😄 😄 😄
